### DISCUSSION TITLE
Get and set MCACC memory profiles

### DISCUSSION THREAD

**Discussion by Tamas-Toth-ebola** [crowbarz/aiopioneer#33] [Q&A] [answered]
Dear @crowbarz !

Thanks for your very nice HA plugin, while I use it, I'm happy with it, and your continuous developments also.

Currently, I have a question, which I simply could not solve yet, and these are the 'get' and 'set' 'MCACC memory profiles/presets' functions.

If I'm right, there is no dedicated GUI function for it _(yet)_ like the sound mode selector, but in the `https://github.com/crowbarz/aiopioneer/blob/main/aiopioneer/commands.py` I found a `query_dsp_mcacc_memory_query` and a corresponding `set_dsp_mcacc_memory_set` command(s), which seems to be what I'm looking for, but sadly I don't know how to call them, neither from HA _(as there is no related services for such commands)_, nor from/with your plugin's CLI interface from my HA container _(while I can reach it as in my case `aiopioneer 192.168.1.52`)_.

Could you please help me, how could it be possible? I mean changing the MCACC profiles as I have 3 one, one for movie watching named as 'sofa', one for general music listening named as 'normal', and one for my working position named as 'chair'.

Thanks for your answer in advance!

_(I have an SC-1223 device if it is a question.)_

**Comment by pantherale0** [comment]
In theory this is already part of the way there. The only thing missing is the ability to set the profile.

Currently this is queried alongside all the rest of the usual queries that are sent, I think you can get the current profile under `dsp.mcacc_memory_set` under the main zone entity in Home Assistant.

Only thing that I think is needed is the ability to set the profile as a function either in this module or the command could be sent directly from the integration.

For reference, any command starting with `query_` will be requested in a standard configuration. It's up to the AVR to actually respond with something.

**Comment by crowbarz** [comment, marked as answer]
In 0.9.0 of the HA integration, the current MCACC memory set is available an attribute of the `DSP` sensor that is created under the AVR device. It's not currently enabled by default because of the HA database updates that it may generate, but you can enable it manually in Settings > Devices & Services > Integrations > Pioneer AVR > (instance) > Devices > (main AVR device).

There's currently no service in the HA integration to send the `set_dsp_mcacc_memory_set` command (crowbarz/ha-pioneer_async#42), or to send an arbitrary command to the AVR (crowbarz/ha-pioneer_async#41), but both should be fairly straightforward to implement.

From the CLI, you can send any command listed in the `PIONEER_COMMANDS` map just by typing the command's name in the CLI, followed optionally by the command's prefix argument.

`set_dsp_mcacc_memory_set` takes a single prefix digit argument (this is from the Pioneer command reference):

| Command | Function | Response | Parameter |
| -- | -- | -- | -- |
| *MC<CR> | MCACC MEMORY SET | MC*<CR+LF> | *:<br>0:MCACC MEMORY (cyclic)<br>1:MEMORY 1<br>2:MEMORY 2<br>3:MEMORY 3<br>4:MEMORY 4<br>5:MEMORY 5<br>6:MEMORY 6<br>9:MCACC MEMORY (cyclic reverse) |

So you should be able to send `set_dsp_mcacc_memory_set 2` from the CLI to select MCACC memory 2. Note that the CLI is for debug and development only so it's preferable to use the supported interfaces once they are available.

**Reply by Tamas-Toth-ebola** [reply]
Hi @crowbarz !

Thanks for the help, my previous problem was my to fast 'response' on the CLI command... as it give us back lot of initialized info, I thought that response is continuous, so I did not find, where to put the commands.

After the starting 'sequence' I could try the mentioned command, which of course had worked.

I will not use the CLI interface, as you already told, that is basically for debugging, so I'm hoping , that you could extent the HA plugin with this function(s),  as you detailed the corresponding 2 'tasks'.

Lot of thanks for it in advance, if you could do it in the future!
